### PR TITLE
feat: Auto-teleport ED followup 

### LINKED ISSUE #8631: Auto-teleport ED followup
from @hassnian 
>atm the autoteleport will check the transferable balance (balance - ed = transferrable balance ) and then check if that amount is greater than the one needed plus current chain balance

I was thinking of introducing some logic 

>If the account would be terminated, but you could save the funds by teleporting the whole amount -> pass

Enable auto-teleport if there are enough funds on the relay chain - if the account would drop below the ed after auto-teleport, we will teleport the whole amount. 

The user won't technically lose funds and can finish the desired action. 

worst thing that can happen is that you would teleport a max of 0.999 DOT more to the parachain

wdyt?

### PR DESCRIPTION
**Pull request by hassnian**
## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

- [x] Closes #8631

#### Did your issue had any of the "$" label on it?

- [x] Fill up your DOT address: [Payout](https://kodadot.xyz/dot/transfer?target=13QUj3pZyFNfYj4AM336hRdyLQbevj5H3sR4PKmLEXLdwZhh)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [x] My fix has changed **something** on UI; 

## Copilot Summary
copilot:summary

copilot:poem

### CODE REVIEW

File: composables/autoTeleport/useAutoTeleportTransitionDetails.ts [hunk 1]
```diff
@@ -157,9 +179,9 @@ export default function (
     return sourceChainProperties?.tokenSymbol === chainSymbol.value
   })
 
-  const fetchTeleportFee = computed(
+  const fetchTeleportFee = computed<boolean>(
```
Comment (roiLeo) [inline composables/autoTeleport/useAutoTeleportTransitionDetails.ts:182]: confusing computed name, maybe use something like `hasTeleportFee` or `hasFee`

**Review by roiLeo — APPROVED** [review]
otherwise lgtm

File: composables/autoTeleport/useAutoTeleportTransitionDetails.ts (hunk 1 above)
Comment (hassnian) [reply composables/autoTeleport/useAutoTeleportTransitionDetails.ts]: changed it to canGetTeleportFee not sure if any better , don't think hasTeleportFee or hasFee  explain well what that computed does wdyt

File: composables/autoTeleport/useAutoTeleportTransitionDetails.ts [hunk 2]
```diff
@@ -122,11 +122,14 @@ export default function (
...
   const richestChainBalance = computed(() =>
     richestChain.value
-      ? Number(sourceChainsBalances.value[richestChain.value]) -
-        richestChainExistentialDeposit.value
+      ? Number(sourceChainsBalances.value[richestChain.value])
       : 0,
   )
 
+  const transferableRichestChainBalance = computed(
+    () => richestChainBalance.value - richestChainExistentialDeposit.value,
```
Comment (daiagi) [inline composables/autoTeleport/useAutoTeleportTransitionDetails.ts:130]: could be < 0

```suggestion

Math.max(richestChainBalance.value - richestChainExistentialDeposit.value, 0)

```

File: composables/autoTeleport/useAutoTeleportTransitionDetails.ts [hunk 3]
```diff
@@ -135,17 +138,36 @@ export default function (
     return bufferFee === 0 ? amountFee : bufferFee
   })
 
-  const amountToTeleport = computed(
+  const onlyRequiredAmountToTeleport = computed(
```
Comment (daiagi) [inline composables/autoTeleport/useAutoTeleportTransitionDetails.ts:141]: like roileo, i also don't undersatnd the name
if you can't name it well, add a comment to explain the significance please

File: composables/autoTeleport/useAutoTeleportTransitionDetails.ts [hunk 4]
```diff
@@ -135,17 +138,36 @@ export default function (
...
+      : onlyRequiredAmountToTeleport.value,
+  )
+
+  const shouldTeleportAllBalance = computed<boolean>(
+    () =>
+      Boolean(richestChain.value) &&
+      !hasEnoughInCurrentChain.value &&
+      richestChainBalance.value - onlyRequiredAmountToTeleport.value <=
+        richestChainExistentialDeposit.value,
   )
```
Comment (daiagi) [inline composables/autoTeleport/useAutoTeleportTransitionDetails.ts:160]: that's a dense one
could you break it up to intermidiate consts with explantory names?

File: composables/autoTeleport/useAutoTeleportTransitionDetails.ts [hunk 5]
```diff
@@ -135,17 +138,36 @@ export default function (
...
 
+  const hasEnoughInRichestChain = computed(() => {
+    const balance =
+      (shouldTeleportAllBalance.value
+        ? richestChainBalance.value
+        : transferableRichestChainBalance.value) || 0
```
Comment (daiagi) [inline composables/autoTeleport/useAutoTeleportTransitionDetails.ts:166]: you would not need || 0 if you cap transferableRichestChainBalance to be minimun 0

File: composables/autoTeleport/useAutoTeleportTransitionDetails.ts [hunk 6]
```diff
@@ -208,8 +230,8 @@ export default function (
     ])
   }
 
-  watch(fetchTeleportFee, async () => {
-    if (fetchTeleportFee.value) {
+  watch(canGetTeleportFee, async () => {
+    if (canGetTeleportFee.value) {
       fetched.value.teleportTxFee = false
       const fee = await getTeleportTransactionFee()
       teleportTxFee.value = Number(fee) || 0
```
Comment (daiagi) [inline composables/autoTeleport/useAutoTeleportTransitionDetails.ts:237]: fetched is an odd name for a ref
plus it looks like it should be `reactive`

**Review by daiagi — CHANGES_REQUESTED** [review]
and on a general not
make sure you work with native amounts and not floting point formatted amounts to avoid all sort of rounding and comparision  issues

**Comment by hassnian** [comment]
@daiagi fixed all comments, let me know if any still need changes

**Review by daiagi — APPROVED** [review]
lgtm after @prury confirms

**Comment by prury** [comment]
let's see if i got it right, i've setup the following test: 

first account had 1.2 DOT on relay
second account had an NFT to be bought for 0.4 DOT on AssetHub ( and had no funds on AssetHub) 
first account did auto-teleport and had to teleport all funds from relay since it would drop below ED

whole process:

https://github.com/kodadot/nft-gallery/assets/36627808/fb67d37a-a109-43fa-9a6c-aa35d0340469

transactions(teleport,buy)
https://polkadot.subscan.io/tx/0xfd4cc7256a6c72a4a4f3ac5fb5a8e8309e3c5bbaaa5567390234fb5ace97a5da
https://assethub-polkadot.subscan.io/tx/0x03bd8ab1d5effb0185af38b4603dbe2631d01060a52ecf4a8a0d6242a2252157

**Comment by yangwao** [comment]
Thanks!
pay 30 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $30 @ 8.64 USD/DOT ~ 3.472 $DOT
🧗 [13QUj3pZyFNfYj4AM336hRdyLQbevj5H3sR4PKmLEXLdwZhh](https://polkadot.subscan.io/account/13QUj3pZyFNfYj4AM336hRdyLQbevj5H3sR4PKmLEXLdwZhh)
🔗 [0x39ce41c2b28b41caa12d71a0e1864ca0c435c7a57c17cf82fa03f538ea3ff6bd](https://polkadot.subscan.io/extrinsic/0x39ce41c2b28b41caa12d71a0e1864ca0c435c7a57c17cf82fa03f538ea3ff6bd)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**